### PR TITLE
Fixed #2584 - better error handling with Hash() (hash.Hash256, error)

### LINKED ISSUE #2584: better error handling with Hash() (hash.Hash256, error)
**What would you like to be added**:
hash of envelope/sealed envelope only returns hash.Hash256, hiding any error that might occur during the process. as a result, panic is used (instead of an all-zero hash value) to indicate error

return (hash.Hash256, error) could mitigate this issue and remove unnecessary panic

**Why is this needed**:
better code

### PR DESCRIPTION
**Pull request by ashishsnigam**
Fixed #2584 

1. Updated function in file /action/sealedenvelope.go to return error also. New signature will be func (sealed *SealedEnvelope) Hash() (hash.Hash256, error)
2. Updated calling of this function as per new signature across entire source code
3. In few places, error of hash function is ignored for brevity

### CODE REVIEW

File: actpool/actpool.go [hunk 1]
```diff
@@ -325,7 +328,8 @@ func (ap *actPool) validate(ctx context.Context, selp action.SealedEnvelope) err
 		return errors.Wrap(action.ErrAddress, "action source address is blacklisted")
 	}
 	// if already validated
-	if _, ok := ap.allActions[selp.Hash()]; ok {
+	selpHash, err := selp.Hash()
```
Comment (CoderZhi) [inline actpool/actpool.go:331]: error handling

File: actpool/actpool.go [hunk 2]
```diff
@@ -460,11 +464,12 @@ func (ap *actPool) removeInvalidActs(acts []action.SealedEnvelope) {
 // deleteAccountDestinationActions just for destination map
 func (ap *actPool) deleteAccountDestinationActions(acts ...action.SealedEnvelope) {
 	for _, act := range acts {
+		hash, _ := act.Hash()
```
Comment (CoderZhi) [inline actpool/actpool.go:467]: error handling

File: api/api.go [hunk 3]
```diff
@@ -1411,7 +1420,7 @@ func (api *Server) actionsInBlock(blk *block.Block, start, count uint64) []*iote
 	}
 	for i := start; i < lastAction; i++ {
 		selp := blk.Actions[i]
-		actHash := selp.Hash()
+		actHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline api/api.go:1423]: error handling

File: api/api.go [hunk 4]
```diff
@@ -1436,7 +1445,7 @@ func (api *Server) reverseActionsInBlock(blk *block.Block, reverseStart, count u
 	for i := reverseStart; i < uint64(len(blk.Actions)) && i < reverseStart+count; i++ {
 		ri := uint64(len(blk.Actions)) - 1 - i
 		selp := blk.Actions[ri]
-		actHash := selp.Hash()
+		actHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline api/api.go:1448]: ditto

File: blockchain/block/block.go [hunk 5]
```diff
@@ -158,7 +158,7 @@ func (b *Block) TransactionLog() *BlkTransactionLog {
 func (b *Block) ActionHashs() []string {
 	actHash := make([]string, len(b.Actions))
 	for i := range b.Actions {
-		h := b.Actions[i].Hash()
+		h, _ := b.Actions[i].Hash()
```
Comment (CoderZhi) [inline blockchain/block/block.go:161]: ditto

File: blockchain/block/utils.go [hunk 6]
```diff
@@ -20,7 +20,8 @@ import (
 func calculateTxRoot(acts []action.SealedEnvelope) hash.Hash256 {
 	h := make([]hash.Hash256, 0, len(acts))
 	for _, act := range acts {
-		h = append(h, act.Hash())
+		actHash, _ := act.Hash()
```
Comment (CoderZhi) [inline blockchain/block/utils.go:23]: ditto

File: blockchain/blockdao/blockdao.go [hunk 7]
```diff
@@ -277,7 +277,8 @@ func (dao *blockDAO) GetActionByActionHash(h hash.Hash256, height uint64) (actio
 		return action.SealedEnvelope{}, 0, err
 	}
 	for i, act := range blk.Actions {
-		if act.Hash() == h {
+		actHash, _ := act.Hash()
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go:280]: ditto

File: blockindex/indexer.go [hunk 8]
```diff
@@ -162,7 +162,7 @@ func (x *blockIndexer) DeleteTipBlock(blk *block.Block) error {
 
 	// delete action index
 	for _, selp := range blk.Actions {
-		actHash := selp.Hash()
+		actHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline blockindex/indexer.go:165]: ditto

File: blockindex/indexer.go [hunk 9]
```diff
@@ -322,7 +322,7 @@ func (x *blockIndexer) putBlock(blk *block.Block) error {
 	}
 	// index actions in the block
 	for _, selp := range blk.Actions {
-		actHash := selp.Hash()
+		actHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline blockindex/indexer.go:325]: ditto

File: state/factory/workingset.go [hunk 10]
```diff
@@ -159,11 +162,12 @@ func (ws *workingSet) runAction(
 	}
 	for _, actionHandler := range reg.All() {
 		receipt, err := actionHandler.Handle(ctx, elp.Action(), ws)
+		elpHash, _ := elp.Hash()
```
Comment (CoderZhi) [inline state/factory/workingset.go:165]: ditto

File: state/factory/workingset.go [hunk 11]
```diff
@@ -421,7 +425,8 @@ func (ws *workingSet) pickAndRunActions(
 				actionIterator.PopAccount()
 				continue
 			default:
-				return nil, errors.Wrapf(err, "Failed to update state changes for selp %x", nextAction.Hash())
+				nextActionHash, _ := nextAction.Hash()
```
Comment (CoderZhi) [inline state/factory/workingset.go:428]: ditto

File: tools/util/injectorutil.go [hunk 12]
```diff
@@ -351,7 +351,8 @@ func DeployContract(
 	log.L().Info("Created signed execution")
 
 	injectExecution(selp, execution, client, retryNum, retryInterval)
-	return selp.Hash(), nil
+	selpHash, err := selp.Hash()
+	return selpHash, nil
```
Comment (CoderZhi) [inline tools/util/injectorutil.go:355]: return error

File: tools/util/injectorutil.go [hunk 13]
```diff
@@ -383,7 +384,8 @@ func injectTransfer(
 	}, bo); err != nil {
 		log.L().Error("Failed to inject transfer", zap.Error(err))
 	} else if pendingActionMap != nil {
-		pendingActionMap.Add(selp.Hash(), 1)
+		selpHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline tools/util/injectorutil.go:387]: log.L().Fatal

File: tools/util/injectorutil.go [hunk 14]
```diff
@@ -416,7 +418,8 @@ func injectExecInteraction(
 	injectExecution(selp, execution, c, retryNum, retryInterval)
 
 	if pendingActionMap != nil {
-		pendingActionMap.Add(selp.Hash(), 1)
+		selpHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline tools/util/injectorutil.go:421]: ditto

File: tools/util/injectorutil.go [hunk 15]
```diff
@@ -485,7 +488,8 @@ func injectStake(
 	}, bo); err != nil {
 		log.L().Error("Failed to inject stake", zap.Error(err))
 	} else if pendingActionMap != nil {
-		pendingActionMap.Add(selp.Hash(), 1)
+		selpHash, _ := selp.Hash()
```
Comment (CoderZhi) [inline tools/util/injectorutil.go:491]: ditto

File: action/sealedenvelope.go [hunk 16]
```diff
@@ -38,22 +38,22 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
...
+			return hash.ZeroHash256, err
 		}
-		return h
+		return h, nil
 	case iotextypes.Encoding_IOTEX_PROTOBUF:
-		return hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto())))
+		return hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto()))), nil
 	}
-	panic("unknown encoding type")
+	return hash.ZeroHash256, errors.Errorf("unknown encoding type %s", sealed.encoding)
```
Comment (dustinxie) [inline action/sealedenvelope.go:56]: %s -> %v

File: actpool/actpool.go (hunk 1 above)
Comment (ashishsnigam) [reply actpool/actpool.go:331]: Got it. Done.

File: actpool/actpool.go (hunk 2 above)
Comment (ashishsnigam) [reply actpool/actpool.go:467]: putting break statement there.

File: api/api_test.go [hunk 17]
```diff
@@ -2130,7 +2130,8 @@ func addTestingBlocks(bc blockchain.Blockchain, ap actpool.ActPool) error {
 	if err != nil {
 		return err
 	}
-	implicitLogs[tsf.Hash()] = block.NewTransactionLog(tsf.Hash(),
+	tsfHash, err := tsf.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2133]: check error

File: api/api_test.go [hunk 18]
```diff
@@ -2162,15 +2163,17 @@ func addTestingBlocks(bc blockchain.Blockchain, ap actpool.ActPool) error {
 		if err := ap.Add(ctx, selp); err != nil {
 			return err
 		}
-		implicitLogs[selp.Hash()] = block.NewTransactionLog(selp.Hash(),
+		selpHash, err := selp.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2166]: check error

File: api/api_test.go [hunk 19]
```diff
@@ -2162,15 +2163,17 @@ func addTestingBlocks(bc blockchain.Blockchain, ap actpool.ActPool) error {
...
+		implicitLogs[selpHash] = block.NewTransactionLog(selpHash,
 			[]*block.TokenTxRecord{block.NewTokenTxRecord(iotextypes.TransactionLogType_NATIVE_TRANSFER, "1", addr3, recipient)},
 		)
 	}
 	selp, err := action.SignedTransfer(addr3, priKey3, uint64(5), big.NewInt(2), []byte{}, testutil.TestGasLimit, big.NewInt(testutil.TestGasPriceInt64))
 	if err != nil {
 		return err
 	}
-	implicitLogs[selp.Hash()] = block.NewTransactionLog(selp.Hash(),
+	selpHash, err := selp.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2175]: check error

File: api/api_test.go [hunk 20]
```diff
@@ -2181,8 +2184,9 @@ func addTestingBlocks(bc blockchain.Blockchain, ap actpool.ActPool) error {
 	if err != nil {
 		return err
 	}
-	implicitLogs[execution1.Hash()] = block.NewTransactionLog(
-		execution1.Hash(),
+	execution1Hash, err := execution1.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2187]: check error

File: api/api_test.go [hunk 21]
```diff
@@ -2240,8 +2252,12 @@ func addTestingBlocks(bc blockchain.Blockchain, ap actpool.ActPool) error {
 	if err != nil {
 		return err
 	}
-	implicitLogs[execution1.Hash()] = block.NewTransactionLog(
-		execution1.Hash(),
+	execution1Hash, err = execution1.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2255]: check error

File: api/api_test.go [hunk 22]
```diff
@@ -2287,7 +2307,8 @@ func deployContract(svr *Server, key crypto.PrivateKey, nonce, height uint64, co
 	// get deployed contract address
 	var contract string
 	if svr.dao != nil {
-		r, err := svr.dao.GetReceiptByActionHash(ex1.Hash(), height+1)
+		ex1Hash, err := ex1.Hash()
```
Comment (dustinxie) [inline api/api_test.go:2310]: check error

File: blockchain/block/block_test.go [hunk 23]
```diff
@@ -67,7 +67,7 @@ func TestMerkle(t *testing.T) {
 
 	hashes := block.ActionHashs()
 	for i := range hashes {
-		h := actions[i].Hash()
+		h, _ := actions[i].Hash()
```
Comment (dustinxie) [inline blockchain/block/block_test.go:70]: h, err =
check error

File: blockchain/integrity/integrity_test.go [hunk 24]
```diff
@@ -321,7 +328,7 @@ func addTestingTsfBlocks(cfg config.Config, bc blockchain.Blockchain, dao blockd
 	if err := ap.Add(context.Background(), ex1); err != nil {
 		return err
 	}
-	deployHash = ex1.Hash()
+	deployHash, err = ex1.Hash()
```
Comment (dustinxie) [inline blockchain/integrity/integrity_test.go:331]: check error

File: blockchain/integrity/integrity_test.go [hunk 25]
```diff
@@ -659,7 +667,8 @@ func addTestingGetBlockHash(t *testing.T, hawaiiHeight uint64, bc blockchain.Blo
 		if err := bc.CommitBlock(blk); err != nil {
 			return hash.ZeroHash256, err
 		}
-		return ex1.Hash(), nil
+		ex1Hash, err := ex1.Hash()
+		return ex1Hash, nil
```
Comment (dustinxie) [inline blockchain/integrity/integrity_test.go:671]: return err

File: actpool/actpool.go (hunk 2 above)
Comment (ashishsnigam) [reply actpool/actpool.go:467]: Done

File: api/api.go (hunk 3 above)
Comment (ashishsnigam) [reply api/api.go:1423]: Done

File: e2etest/local_transfer_test.go [hunk 26]
```diff
@@ -394,7 +394,8 @@ func TestLocalTransfer(t *testing.T) {
 			var selp action.SealedEnvelope
 			err := backoff.Retry(func() error {
 				var err error
-				selp, err = as.GetActionByActionHash(tsf.Hash())
+				tsfHash, _ := tsf.Hash()
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:397]: tsfHash, err =
check error

File: api/api.go (hunk 4 above)
Comment (ashishsnigam) [reply api/api.go:1448]: done

File: e2etest/local_transfer_test.go [hunk 27]
```diff
@@ -394,7 +394,8 @@ func TestLocalTransfer(t *testing.T) {
...
+				selp, err = as.GetActionByActionHash(tsfHash)
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:398]: check error

File: e2etest/local_transfer_test.go [hunk 28]
```diff
@@ -442,11 +443,14 @@ func TestLocalTransfer(t *testing.T) {
 			//Wait long enough to make sure the failed transfer does not exit in either action pool or blockchain
 			err := backoff.Retry(func() error {
 				var err error
-				_, err = ap.GetActionByHash(tsf.Hash())
+				tsfHash, _ := tsf.Hash()
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:446]: tsfHash, err =
check error

File: blockchain/block/block.go (hunk 5 above)
Comment (ashishsnigam) [reply blockchain/block/block.go:161]: done

File: e2etest/local_transfer_test.go [hunk 29]
```diff
@@ -459,11 +463,14 @@ func TestLocalTransfer(t *testing.T) {
 			//Need to wait long enough to make sure the pending transfer is not minted, only stay in action pool
 			err := backoff.Retry(func() error {
 				var err error
-				_, err = ap.GetActionByHash(tsf.Hash())
+				tsfHash, _ := tsf.Hash()
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:466]: tsfHash, err =
check error

File: e2etest/native_staking_test.go [hunk 30]
```diff
@@ -165,7 +169,8 @@ func TestNativeStaking(t *testing.T) {
 		require.Equal(hash.BytesToHash256(cand1Addr.Bytes()), logs[0].Topics[2])
 		voter1BucketIndex := byteutil.BytesToUint64BigEndian(logs[0].Topics[1][24:])
 
-		r1, err = dao.GetReceiptByActionHash(cs2.Hash(), 3)
+		cs2Hash, err := cs2.Hash()
```
Comment (dustinxie) [inline e2etest/native_staking_test.go:172]: check error

File: e2etest/rewarding_test.go [hunk 31]
```diff
@@ -513,7 +513,8 @@ func injectClaim(
 	}
 
 	if err == nil {
-		pendingClaimActions[selp.Hash()] = expectedSuccess
+		selpHash, _ := selp.Hash()
```
Comment (dustinxie) [inline e2etest/rewarding_test.go:516]: selpHash, err =
check error

File: blockchain/block/utils.go (hunk 6 above)
Comment (ashishsnigam) [reply blockchain/block/utils.go:23]: done

File: blockchain/blockdao/blockdao.go (hunk 7 above)
Comment (ashishsnigam) [reply blockchain/blockdao/blockdao.go:280]: done

File: blockindex/indexer.go (hunk 8 above)
Comment (ashishsnigam) [reply blockindex/indexer.go:165]: done

File: blockindex/indexer.go (hunk 9 above)
Comment (ashishsnigam) [reply blockindex/indexer.go:325]: done

File: state/factory/workingset.go (hunk 10 above)
Comment (ashishsnigam) [reply state/factory/workingset.go:165]: done

File: state/factory/workingset.go (hunk 11 above)
Comment (ashishsnigam) [reply state/factory/workingset.go:428]: done

File: tools/util/injectorutil.go (hunk 12 above)
Comment (ashishsnigam) [reply tools/util/injectorutil.go:355]: done

File: tools/util/injectorutil.go (hunk 13 above)
Comment (ashishsnigam) [reply tools/util/injectorutil.go:387]: done

File: action/sealedenvelope.go (hunk 16 above)
Comment (dustinxie) [reply action/sealedenvelope.go:56]: and put this into a `default:` clause

File: tools/util/injectorutil.go (hunk 14 above)
Comment (ashishsnigam) [reply tools/util/injectorutil.go:421]: done

File: tools/util/injectorutil.go (hunk 15 above)
Comment (ashishsnigam) [reply tools/util/injectorutil.go:491]: done

File: action/sealedenvelope.go (hunk 16 above)
Comment (ashishsnigam) [reply action/sealedenvelope.go:56]: Both points done

File: api/api_test.go (hunk 17 above)
Comment (ashishsnigam) [reply api/api_test.go:2133]: done

File: action/sealedenvelope.go (hunk 16 above)
Comment (dustinxie) [reply action/sealedenvelope.go:56]: also line 121 below: %s -> %v

File: api/api_test.go (hunk 18 above)
Comment (ashishsnigam) [reply api/api_test.go:2166]: done

File: api/api_test.go (hunk 19 above)
Comment (ashishsnigam) [reply api/api_test.go:2175]: done

File: api/api_test.go (hunk 20 above)
Comment (ashishsnigam) [reply api/api_test.go:2187]: done

File: api/api_test.go (hunk 21 above)
Comment (ashishsnigam) [reply api/api_test.go:2255]: done

File: api/api_test.go (hunk 22 above)
Comment (ashishsnigam) [reply api/api_test.go:2310]: done

File: blockchain/block/block_test.go (hunk 23 above)
Comment (ashishsnigam) [reply blockchain/block/block_test.go:70]: done

File: blockchain/integrity/integrity_test.go (hunk 24 above)
Comment (ashishsnigam) [reply blockchain/integrity/integrity_test.go:331]: done

File: blockchain/integrity/integrity_test.go (hunk 25 above)
Comment (ashishsnigam) [reply blockchain/integrity/integrity_test.go:671]: done

File: e2etest/local_transfer_test.go (hunk 26 above)
Comment (ashishsnigam) [reply e2etest/local_transfer_test.go:397]: done

File: e2etest/local_transfer_test.go (hunk 27 above)
Comment (ashishsnigam) [reply e2etest/local_transfer_test.go:398]: done

File: e2etest/local_transfer_test.go (hunk 28 above)
Comment (ashishsnigam) [reply e2etest/local_transfer_test.go:446]: done

File: e2etest/local_transfer_test.go (hunk 29 above)
Comment (ashishsnigam) [reply e2etest/local_transfer_test.go:466]: done

File: e2etest/native_staking_test.go (hunk 30 above)
Comment (ashishsnigam) [reply e2etest/native_staking_test.go:172]: done

File: e2etest/rewarding_test.go (hunk 31 above)
Comment (ashishsnigam) [reply e2etest/rewarding_test.go:516]: done

**Comment by dustinxie** [comment]
1. looks like CLA did not pass b/c you added commits under another github ID, can you try move all commits to `ashishsnigam` and push -f again? or close this PR and reopen a new one with all commits under `ashishsnigam`
2. pls check more comments in `sealedenvelope.go`

File: tools/util/injectorutil.go (hunk 12 above)
Comment (ashishsnigam) [reply tools/util/injectorutil.go:358]: Changes done, somehow github is not showing here as Outdated. Please see file in expanded mode.

**Comment by ashishsnigam** [comment]
@dustinxie @CoderZhi All review comments incorporated. Please have a look. Thanks.

**Comment by ashishsnigam** [comment]
Opening a new PR.

**Pull request closed without merging**